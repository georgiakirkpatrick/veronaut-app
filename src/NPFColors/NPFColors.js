import React from 'react'
import '@fortawesome/react-fontawesome'
import FormPage from '../FormPage/FormPage'
import FormButton from '../FormButton/FormButton'
import FormDropdown from '../FormDropdown/FormDropdown'
import FormFieldset from '../FormFieldset/FormFieldset'
import FormPromptWithSub from '../FormPromptWithSub/FormPromptWithSub'
import NPFFooter from '../NPFFooter/NPFFooter';
import FormTextInput from '../FormTextInput/FormTextInput'
import formData from '../FORM_DATA'

const NPFColors = props => {

    const {
        colorFieldsets,
        currentPage,
        setColorFieldsets,
        setPage,
    } = props

    const addColor = () => {
        setColorFieldsets(
            [
                ...colorFieldsets, 
                {
                    name: '', 
                    descriptionId: 0, 
                    swatchUrl: '',
                    imageUrls: ['']
                }
            ]
        )
    }

    const colorsChangeInput = (index, event) => {
        const values = [...colorFieldsets]
        console.log('values', values)
        values[index][event.target.name] = event.target.value
        console.log('values[index][event.target.name]', values[index][event.target.name])

        setColorFieldsets(values)
        console.log('values', values)

    }

    const makeColorOptions = () => {
        const defaultOption = {
            id: 0,
            text: 'Select a color',
            value: 0
        }
        
        const colors = formData.colors.map(color => (
            {
                id: color.id,
                text: color.text,
                value: color.id
            }
        ))

        return [
            defaultOption,
            ...colors
        ]
    }

    const nextButton = () => {
        const missingFields = []

        const requiredFields = [
            'descriptionId'
        ]

        colorFieldsets.length > 1 && requiredFields.push('name')

        colorFieldsets.forEach(fieldset => {
            requiredFields.forEach(field => {
                if (fieldset[field] === 0 || fieldset[field] === '') {
                    missingFields.push(field)
                }
            })
        })

        if (missingFields.length >= 1) {
            alert(`Please enter a description for each color option.  Remove any unneeded color fields by clicking 'remove'.`)
        } else if (missingFields.length === 0) {
            setPage(currentPage + 1)
        }
    }
    
    // const nextButton = () => {setPage(currentPage + 1)}
    
    const removeColor = (index) => {
        const values = [...colorFieldsets]
        values.splice(index, 1)
        setColorFieldsets(values)
    }

    return (
        <div id='colors'>
            <FormPage title='Colors'>
                <div id='color-inputs'>
                    <FormPromptWithSub 
                        prompt='Enter all color options listed on the product webpage'
                        promptSubtitle='If the product has multiple colors, what is the dominant color?'
                    />
                    {colorFieldsets.map((colorFieldset, index) => (
                        <FormFieldset key={index} className='NewProductForm__fieldset'>
                            <button 
                                className='NewProductForm__remove' 
                                type='button' 
                                onClick={() => removeColor(index)}
                            >
                                REMOVE
                            </button>
                            <FormTextInput 
                                id={'color-name' + index}
                                name='name'
                                prompt='Color name'
                                currentValue={colorFieldset.name} 
                                handleChange={event => {
                                    colorsChangeInput(index, event)
                                }}
                            />
                            <FormDropdown 
                                id={'color-description' + index} 
                                name='descriptionId'
                                prompt='Color description'
                                // 'If the product has multiple colors, what is the dominant color?'
                                options={makeColorOptions()}
                                currentValue={colorFieldset.descriptionId} 
                                handleChange={event => {
                                    colorsChangeInput(index, event)
                                }}
                            />
                        </ FormFieldset>
                    ))}
                </div>
    
                <FormButton 
                    buttonText='ADD A COLOR' 
                    handleClick={() => addColor() } 
                />                
            </FormPage>
            <NPFFooter
                buttons='prevNext'
                previousButton={() => setPage(currentPage - 1)} 
                nextButton={() => {
                    nextButton()
                }}
            />
        </div>
    )
}

NPFColors.defaultProps = {
    currentPage: 0,
    colorFieldsets: [],
    setColorFieldsets: () => {},
    setPage: () => {},
}

export default NPFColors