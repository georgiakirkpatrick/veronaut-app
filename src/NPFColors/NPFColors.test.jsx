import React from 'react'
import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import NPFColors from './NPFColors'

describe('NPFColors', () => {
  it ('renders without crashing', () => {
    render(<NPFColors />)
    screen.debug()
  })

  it ('render the UI as expected', () => {
    const tree = render(<NPFColors />)
    expect(tree).toMatchSnapshot()
  })
})