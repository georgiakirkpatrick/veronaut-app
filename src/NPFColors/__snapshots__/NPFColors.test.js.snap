// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NPFColors render the UI as expected 1`] = `
<div
  id="colors"
>
  <section
    className="FormPage"
  >
    <h1>
      Colors
    </h1>
    <form
      autoComplete="off"
      id="form-page"
    >
      <div
        className="FormPromptWithSub"
      >
        <p
          className="FormPromptWithSub__prompt"
        >
          Enter all color options listed on the product webpage
        </p>
        <p
          className="FormPromptWithSub__subtitle"
        >
          If a product color option contains multiple colors, what is the dominant color?
        </p>
      </div>
      <button
        className="FormButton"
        onClick={[Function]}
        type="button"
      >
        Add a color option
      </button>
    </form>
    <div>
      
    </div>
  </section>
  <footer
    className="NPFFooter"
  >
    <button
      className="NPFFooter__previous"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-left fa-w-8 NPFFooter__left-arrow"
        data-icon="angle-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <span>
        PREVIOUS
      </span>
    </button>
    <button
      className="NPFFooter__next"
      onClick={[Function]}
      type="button"
    >
      <span>
        NEXT
      </span>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-right fa-w-8 NPFFooter__right-arrow"
        data-icon="angle-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </footer>
</div>
`;
